### DISCUSSION TITLE
why not build a simpler version?

### DISCUSSION THREAD

**Discussion by erwinooghe** [jamulussoftware/jamulus#2157] [Ideas]
Hi,

I sing in a choir of about 30 people and we are testing several possibilities to sing online. Jamulus is great, but, in fact, a simpler version would do for us (and probably many other bands):

1/ Currently, Jamulus organizes the collection and synchronization of the different instruments (voices in our case) and returns the synchronized signal to everyone

2/ Unfortunately, delay is an issue (depending on many factors, I know, but with 30 people it will be very hard to get it right and we want to avoid people dropping out)

3/ Yet, there might be a simple solution: suppose the choir leader can play a song (say, an mp3). Everyone sings along (with mp3 in headphones), and Jamulus again organizes the collection and synchronization of the voices, but returns the synchronized signal to the choir leader (who can adjust volume/balance of the different voices etc). 

4/ The choir leader hears it with some delay, but that's perfectly fine ... and if she can record it, even better, than everyone can hear it afterwards and we can have a discussion.

Unfortunately I am a coding dummy, but I am convinced that some features of Jamulus (the part that collects and synchronizes the instruments) can probably be used and adjusted to build, say, a Jamulus-light. Alternatively, some of you may know software that do the trick and that are available for free? (I did not encounter it up to now)

Thanks for any advice, help, suggestions, ...

**Comment by ann0see** [comment]
Isn’t that already possible by muting everyone but the choir leader?

However, singing with ~ 30 people should be possible if you focus on the leader. OFC. that needs practice but for us that worked well. There are even choirs with ~ 50 PPL which can sing together.

**Reply by erwinooghe** [reply]
That's a great suggestion indeed (with hindsight, slightly silly that I did not think of it myself). To be sure I understand it correctly, the choir leader mutes herself in the mix, plays an mp3 (through mic or as instrument if possible), choir members mute everyone except choir leader (so, we solo her), we sing along with her mp3, and she hears us singing (with some delay, but synchronized). Is there an easy way for her to record it directly? I did not see it in Jamulus, but maybe there are simple workarounds? Thanks again, that really helped me to understand the possibilities of Jamulus!

**Reply by ann0see** [reply]
Yeah, that’s a way. 

Concerning the recording/playback: it is certainly possible with JACK (my personal preference due to low delay and good integration on Linux Audio Production Tools), Voicemeeter (probably not that bad on Windows; however I find it confusing). Have a look at the document linked on https://jamulus.io/wiki/Tips-Tricks-More

**Reply by gene96817** [reply]
I read the original message as struggling with two issues.
1- learning to sing with delay. Note that the delay is different for each singer. This is a good skill to develop.
2- singing against an mp3 (#3) and then mixing all the tracks together. For this requirement the Jamulus-lite is Audacity.
     Personally, I consider this a recording project. Everyone creates a recording against the mp3. Then everything is brought together for the complete mix. This eliminates the need to listen to each other in the creation of music.

**Reply by erwinooghe** [reply]
(i copy my reply here, seems that replying by e-mail arrives elsewhere; sorry, also new to github)

Hi
Thanks for your response. 2 is spot on: we are indeed looking for a recording project, but one that – in contrast to what you write – works automatically (you hear an mp3, you sing along, and everything is automatically synchronized), does it exist? Doing it manually is somewhat of a hassle – we tried to do it in bandlab (as you can work simultaneously), but it did not work out.

Erwin

**Reply by gene96817** [reply]
Note that you also have a choice of using an mp4. With an mp4, the director can provide some direction, especially for fermatas or retards, etc. 

For the choruses I know, it is common to have a count to establish the tempo and a synchronization signal, a clap. Then the clap is used to align the tracks. I haven't heard of any automatic tools to align all the tracks. Note that once you get to this step in Audacity, you are ready to slip down the rabbit hole on audio editing. You can get seduced into spending a lot of hours to make the recording sound better.

**Comment by erwinooghe** [comment]
Great advice, thanks!

**Comment by gilgongo** [comment]
Is this how https://sub.live/ works ("Backing track with 'Sendahead' for synchronised output.")?

**Reply by erwinooghe** [reply]
Hi, thanks for the suggestion, I didn't know it (but it seems pretty new). For sure: interesting that it has a backing track - option, not sure yet what the "lookahead"-delay can be used for. Also not sure yet how many people can enter a room and whether recording is possible (not in sublive, but maybe via JACK, as suggested in FAQ of sublive) ... I will check it, thanks again!

**Reply by erwinooghe** [reply]
Hi

FYI: although sublive is very nice for several reasons (nice simple layout/use, no server needed, private rooms with invitation possibility, backtrack available), there seems to be a limit of 5 persons per room ...
[https://www.reddit.com/r/sublive/comments/pwhljr/using_for_a_choir/  ](url) 
... which is not suitable for a choir (except for separate voice groups maybe).

So, I keep on looking for an “automatic synchronized recording tool”, say, choir leader (or someone else) sends a backtrack, everybody sings along, tool collects, synchronizes, and records all voices (with or without backtrack), and choir leader (or anyone else) can play the recording for discussion ... Although it is not singing together in real-time, it avoids tricky delay issues (with large groups) as it does not matter that the recording arrives only after the last note of the last (most delayed) singer ... 
(note that we tried bandlab, but it was not very successful).

**Reply by gilgongo** [reply]
If you find out more about other systems, you might like to add them here: https://en.wikipedia.org/wiki/Comparison_of_Remote_Music_Performance_Software 

(I've also just emailed the maintainer of solocontutti.com about that, which is another system I've just found)

**Comment by gene96817** [comment]
> Is this how https://sub.live/ works ("Backing track with 'Sendahead' for synchronized output.")?

Interesting claims. Anyone interested in doing some test. I am interested in two test configurations with 3-6 musicians.
Test 1: "short near-city test" Honolulu to Seatle, Vancover San Francisco, or San Diego
             yeah, these are the "near-cities" more than 350 km east of Honolulu
Test 2: "middle distance test - Honolulu to Chicago or Toronto

**Comment by ann0see** [comment]
Do you know https://jamulus.io/de/wiki/Tips-Tricks-More and https://docs.google.com/document/d/1smcvsxdaaViPQvGMQHmah_6BQeqowhmGSFMHfnlY2FI/edit#heading=h.u12du2nhp6d already? They might help you.

A few thoughts on the general topic you're thinking of. Basically you want 

1. To record the output:

Jamulus has multiple ways to do that. 

* If you're running a server, just [enable recording](https://jamulus.io/wiki/Running-a-Server#recording-directory) and process the files in Audacity or Reaper. You'd play them back via Jamulus (or as mentioned below)
* If you're not running a server, you need to somehow record the audio on your PC (digital world) or in the analog world (phone, ...). This might be done with software (Step 1) or Hardware (Step 2):

1. Software: JACK/Voicemeeter/Audio routing/ASIO/... Not trivial to set up - especially on Windows. However once it runs, you can basically do everything. Suggestion: Find someone - or be the one who sets everything up. (I'd suggest to use JACK on linux/ubuntu studio - maybe even a live disk since IMHO that's the closest way to see the audio flow like it was in the analog world. You'd connect a "DAW" like Ardour, ProTools,... or  just VLC media player to Jamulus, which would then play back or record audio files)
2. Or If you don't want to fiddle along with software, buy an audio interface, connect your phone/recording device to it and use it as hardware in/output. In my opinion the hardware solution is much easier, but not always that flexible.

With both approaches you should be able to do almost everything you need.

2. Unfortunately, delay is an issue

Yes it is; From my experience choirs will get used to it: https://www.youtube.com/watch?v=Z9BJhhi1zEQ https://www.youtube.com/watch?v=m811fZiPyaQ Follow the conductor/piano and don't try to just listen on others. You will suffer from the slow down effect if not everyone has low latency.

My choir once did [live-streaming via 5G](https://github.com/jamulussoftware/jamulus/discussions/1800). This led to overall delay times at around 80 to 100ms. Despite the delay it was possible to sing together and we had a nice concert.

**Reply by erwinooghe** [reply]
Hi
Meanwhile I indeed figured out how to set up a server and how to record it. And yes, I will try to figure out how I can easily playback the files in Jamulus, thanks again for the many suggestions (probably hardware will be the best choice for us given that we want to do limited things, say record, playback, etc during rehearsals).
We are going to test things this weekend. Is there anything known about delay formulas? My (very very wild and uninformed) guess is that the delay of person i, given n singers, is roughly equal to: processing of n people on server + max(ping_k)+ping_i, but probably some of you have info on it?  so, the bottleneck then is the processing of all the signals?

**Reply by dcorson-ticino-com** [reply]
The delay because of processing is small and is completely covered up by the buffering.  One packet is processed while the next is being received.  Note that if the processing takes longer than the buffering Jamulus breaks up and is no longer usable.  For big groups the server needs to have appropriate HW.  But there have been groups of over 100 people successfully singing together. 
The biggest delay component is simply the distance between players/singers and the quality of their Internet connections (jitter).  
There is little or nothing that one can do to influence it.
The best way to overcome the latency is to play a lot using Jamulus and get used to compensating. It happens pretty automatically.

**Reply by erwinooghe** [reply]
Hi, thanks. I run the server on my computer, decent, but not really appropriate HW I guess. Distance should be ok (we almost all live in the same town), but quality might be the bottleneck. (is there a good way to test for jitter? I found some suggestions, but not sure whether I can trust them). In any case, we will test Jamulus first (using my computer as a public server), afterwards we can think of server solutions ...

**Reply by dcorson-ticino-com** [reply]
For a choir of 30 people you shouldn't need anything terribly special.  A server on a RaspberryPI will not cut the mustard, but using your home solution on a desktop you will probably hit Internet bandwidth limits first unless you have a real good (fiber optic) connection.  You will probably find that using a cloud-based solution for the server is best eventually. (Each client has a max bandwidth at best quality stereo of about 1Meg, usually no problem.  But the server has the combined bandwidth of all the clients, that adds up fast.)
What I said about distance may be  little misleading, in reality it is the ping time.  Sometimes the distances Internet traffic goes is hard to believe.  I had a colleague here who had consistently high ping times although less than 10km away.  His provider was apparently sending all traffic through a hub in Germany, effectively adding >700km to the distance.  Changing the provider was the only solution, but not always practical.
Have fun with your choir using Jamulus.  It is a great tool and I have been using it for almost 2 years now.  It has kept me sane. But be realistic, even Jamulus can't evade the laws of physics and the vagarities of Internet routing over which we have no control.

**Reply by ann0see** [reply]
If you already make your server public, you could also try to use a free server in the Choral genre? They should have a good connection . Jamulix and Cantare seem to allow > 30 people

**Reply by ann0see** [reply]
Ok. Did some tests on Windows.

In theory it should be possible to get decent routing by using
JACK and ASIO Bridge. You'd set ASIO Bridge and Jamulus to JackRouter and the system audio to HiFi Cable input. Then you'd do the routing in the JACK Graph settings. 

Probably it would be worth to document the "how" somewhere. But that's another discussion ;-)

**Reply by erwinooghe** [reply]
thanks. I do not understand it fully at this stage, but I will delve into it ...

**Reply by ann0see** [reply]
Yeah, it’s challenging at first since you need to understand what these programs all do and how they Play together. 

Think of ASIO Bridge like a virtual microphone capturing your Windows output and sending it to an ASIO device which is in my case JACK Router. That’s like a virtual mixer with which you can Route audio from application A to application B etc.

Jamulus would also use JACK router as ASIO driver. 

Then in JACK you‘d connect Jamulus and the ASIO Bridge together which results in the audio getting sent to Jamulus. Moreover, you can also connect your normal "System" audio device (interface, microphone,…) via JACK to Jamulus.

**Reply by erwinooghe** [reply]
Thanks for clarifying, i will check the possibilities of asio bridge... I managed to connect my daw to jamulus via jack router which allows me to play a song in jamulus. Probably jack will also allow me to record from jamulus, to be explored tomorrow, best, eo

**Reply by ann0see** [reply]
Great news! Once you’ve connected your DAW basically you‘re set. Load the mp3, connect the DAW output to Jamulus and that’s it. 
😀
Similar for the recording step. 

***

Just a quick note: You can also try the native JACK version of Jamulus which you can find on the release page: https://github.com/jamulussoftware/jamulus/releases/tag/r3_8_1

It‘s probably not that different though.
